aiagents4pharma/talk2knowledgegraphs/utils/extractions/multimodal_pcst.py

```diff
@@ -0,0 +1,292 @@
+    root: int = -1
+    num_clusters: int = 1
+    pruning: str = "gw"
+    verbosity_level: int = 0
+    use_description: bool = False
+
+    def _compute_node_prizes(self,
+                             graph: Data,
+                             query_emb: torch.Tensor,
+                             modality: str) :
+        """
+        Compute the node prizes based on the cosine similarity between the query and nodes.
+
+        Args:
+            graph: The knowledge graph in PyTorch Geometric Data format.
+            query_emb: The query embedding in PyTorch Tensor format. This can be an embedding of
+                a prompt, sequence, or any other feature to be used for the subgraph extraction.
+            modality: The modality to use for the subgraph extraction based on the node type.
+
+        Returns:
+            The prizes of the nodes.
+        """
+        # Convert PyG graph to a DataFrame
+        graph_df = pd.DataFrame({
+            "node_type": graph.node_type,
+            "desc_x": [x.tolist() for x in graph.desc_x],
+            "x": [list(x) for x in graph.x],
+            "score": [0.0 for _ in range(len(graph.node_id))],
+        })
+
+        # Calculate cosine similarity for text features and update the score
+        if self.use_description:
+            graph_df.loc[:, "score"] = torch.nn.CosineSimilarity(dim=-1)(
+                    query_emb,
+                    torch.tensor(list(graph_df.desc_x.values)) # Using textual description features
+                ).tolist()
+        else:
+            graph_df.loc[graph_df["node_type"] == modality,
+                         "score"] = torch.nn.CosineSimilarity(dim=-1)(
+                    query_emb,
+                    torch.tensor(list(graph_df[graph_df["node_type"]== modality].x.values))
+                ).tolist()
+
+        # Set the prizes for nodes based on the similarity scores
+        n_prizes = torch.tensor(graph_df.score.values, dtype=torch.float32)
+        # n_prizes = torch.nn.CosineSimilarity(dim=-1)(query_emb, graph.x)
+        topk = min(self.topk, graph.num_nodes)
+        _, topk_n_indices = torch.topk(n_prizes, topk, largest=True)
+        n_prizes = torch.zeros_like(n_prizes)
+        n_prizes[topk_n_indices] = torch.arange(topk, 0, -1).float()
+
+        return n_prizes
+
+    def _compute_edge_prizes(self,
+                             graph: Data,
+                             prompt_emb: torch.Tensor) :
+        """
+        Compute the node prizes based on the cosine similarity between the query and nodes.
+
+        Args:
+            graph: The knowledge graph in PyTorch Geometric Data format.
+            prompt_emb: The prompt embedding in PyTorch Tensor format.
+
+        Returns:
+            The prizes of the nodes.
+        """
+        # Note that as of now, the edge features are based on textual features
+        # Compute prizes for edges
+        e_prizes = torch.nn.CosineSimilarity(dim=-1)(prompt_emb, graph.edge_attr)

```

Comment:
This also looks like a normalized similarity score that could be computed by a vector database.